File: MaximumProfitinJobScheduling.c
#include<stdio.h>
#include<string.h>
#include<math.h>
#include<stdlib.h>
#include<stdbool.h>
void quicksort(int l, int r, int *a, int *b, int *c)
{
	if (l >= r)	return;
	int i = l, j = r, x[3] = {a[r], b[r], c[r]};
	while (i < j)
	{
		while (i < j && b[i] <= x[1])	++ i;
		a[j] = a[i];
		b[j] = b[i];
		c[j] = c[i];
		while (i < j && b[j] > x[1])	-- j;
		a[i] = a[j];
		b[i] = b[j];
		c[i] = c[j];
	}
	a[i] = x[0];
	b[i] = x[1];
	c[i] = x[2];
	quicksort(l, i - 1, a, b, c);
	quicksort(i + 1, r, a, b, c);
}
int findlast(int y, int *x, int n)
{
	if (y >= x[n])	return n;
	int l = 0, r = n;
	while (l < r)
	{
		if (y < x[(l + r) / 2])
			r = (l + r) / 2;
		else
			l = (l + r) / 2 + 1;
	}
	return l - 1;
}
int jobScheduling(int *startTime, int startTimeSize,
					int *endTime, int endTimeSize,
					int *profit, int profitSize)
{
	int n = profitSize;
	quicksort(0, n - 1, startTime, endTime, profit);
	int a[50001], b[50001];
	a[0] = b[0] = 0;
	for (int i = 0; i < n; ++ i)
	{
		printf("%d\n", i);
		int pos = findlast(*(startTime + i), b, i);
		if (a[pos] + *(profit + i) > a[i])
		{
			a[i + 1] = a[pos] + *(profit + i);
			b[i + 1] = *(endTime + i);
		}
		else
		{
			a[i + 1] = a[i];
			b[i + 1] = b[i];
		}
	}
	return a[n];
}
int main()
{
	FILE *fin = fopen("oo.xx", "r");
	FILE *fout = fopen("xx.oo", "w");
	int n, a[50000], b[50000], c[50000];
	fscanf(fin, "%d", &n);
	for (int i = 0; i < n; ++ i)
		fscanf(fin, "%d,", a + i);
	for (int i = 0; i < n; ++ i)
		fscanf(fin, "%d,", b + i);
	for (int i = 0; i < n; ++ i)
		fscanf(fin, "%d,", c + i);
	int out = obScheduling(a, n, b, n, c, n);
	printf("%d\n", out);
	fprintf(fout, "%d\n", out);
	return 0;
}

File: MaximumProfitinJobScheduling.py
from typing import List
from functools import partial as par
import functools as ft
import sys
import bisect
def F(*z):
    z = [*z]
    z[0] = [z[0]]
    return [*ft.reduce(lambda x, y: map(y, x), z)][0]
FF = lambda *z: [*ft.reduce(lambda x, y: map(y, x), z)]
fyx = lambda f, *x: lambda *y: f(*y, *x)

class Solution:
    def quicksort(l, r, obj, rank):
        if (l >= r):
            return
        i, j, x = l, r, obj[r]
        while (i < j):
            while i < j and obj[i][rank] <= x[rank]:
                i += 1
            obj[j] = obj[i]
            while i < j and obj[j][rank] > x[rank]:
                j -= 1
            obj[i] = obj[j]
        obj[i] = x
        Solution.quicksort(l, i - 1, obj, rank)
        Solution.quicksort(i + 1, r, obj, rank)
    def findpos(x, y):
        n = len(x) - 1
        if (y >= x[-1]):
            return n
        l, r = 0, n - 1
        while 1:
            if y < x[(l + r) // 2]:
                r = (l + r) // 2 
            else:
                l = (l + r) // 2 + 1
            if (y < x[l]):
                return l - 1
    def jobScheduling(
        self, startTime: List[int], endTime: List[int], profit: List[int]
    ) -> int:
        n = len(startTime)
        obj = [*zip(startTime, endTime, profit)]
        obj.sort(key = lambda x: x[1])
#       sys.setrecursionlimit(50001)
#       Solution.quicksort(0, n - 1, obj, 1)
        out = [[0], [0]]
        for s, e, p in obj:
#           pos = Solution.findpos(out[1], s)
            pos = bisect.bisect(out[1], s) - 1
            if (out[0][pos] + p > out[0][-1]):
                out[0].append(out[0][pos] + p)
                out[1].append(e)
        return out[0][-1]

fin = open("oo.xx", "r")
fout = open("xx.oo", "w")

n = int(fin.readline())
startTime = [int(x) for x in fin.readline().split(',')]
endTime = [int(x) for x in fin.readline().split(',')]
profit = [int(x) for x in fin.readline().split(',')]

out = Solution().jobScheduling(startTime, endTime, profit)
print(out)
F(out, str, fout.write)

